datafusion/functions-window/src/nth_value.rs

```diff
@@ -0,0 +1,555 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! `nth_value` window function implementation
+
+use crate::utils::{get_scalar_value_from_args, get_signed_integer};
+
+use std::any::Any;
+use std::cmp::Ordering;
+use std::fmt::Debug;
+use std::ops::Range;
+use std::sync::OnceLock;
+
+use datafusion_common::arrow::array::ArrayRef;
+use datafusion_common::arrow::datatypes::{DataType, Field};
+use datafusion_common::{exec_datafusion_err, exec_err, Result, ScalarValue};
+use datafusion_expr::window_doc_sections::DOC_SECTION_ANALYTICAL;
+use datafusion_expr::window_state::WindowAggState;
+use datafusion_expr::{
+    Documentation, Literal, PartitionEvaluator, ReversedUDWF, Signature, TypeSignature,
+    Volatility, WindowUDFImpl,
+};
+use datafusion_functions_window_common::field;
+use datafusion_functions_window_common::partition::PartitionEvaluatorArgs;
+use field::WindowUDFFieldArgs;
+
+get_or_init_udwf!(
+    First,
+    first_value,
+    "returns the first value in the window frame",
+    NthValue::first
+);
+get_or_init_udwf!(
+    Last,
+    last_value,
+    "returns the last value in the window frame",
+    NthValue::last
+);
+get_or_init_udwf!(
+    NthValue,
+    nth_value,
+    "returns the nth value in the window frame",
+    NthValue::nth
+);
+
+/// Create an expression to represent the `first_value` window function
+///
+pub fn first_value(arg: datafusion_expr::Expr) -> datafusion_expr::Expr {
+    first_value_udwf().call(vec![arg])
+}
+
+/// Create an expression to represent the `last_value` window function
+///
+pub fn last_value(arg: datafusion_expr::Expr) -> datafusion_expr::Expr {
+    last_value_udwf().call(vec![arg])
+}
+
+/// Create an expression to represent the `nth_value` window function
+///
+pub fn nth_value(arg: datafusion_expr::Expr, n: Option<i64>) -> datafusion_expr::Expr {
```

Comment:
The type of `n` is `i64`, not `Option<i64>`.

See the rust docs: [https://docs.rs/datafusion/latest/datafusion/logical_expr/window_function/fn.nth_value.html](https://docs.rs/datafusion/latest/datafusion/logical_expr/window_function/fn.nth_value.html)

Reply:
Also add a roundtrip logical plan test for this API here: https://github.com/apache/datafusion/blob/ac79ef3442e65f6197c7234da9fad964895b9101/datafusion/proto/tests/cases/roundtrip_logical_plan.rs#L943

Reply:
Fixed and added test